### PR TITLE
ti-dmc: Add a new plugin to update some USB-4 docks

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [X] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/ti-dmc/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+The TPS65987D and TPS65988 are Type-C Port Switch / Power Delivery (PD) Controller devices.
+
+## Firmware Format
+
+The daemon will decompress the cabinet archive and extract a firmware blob as a SHA-256+RSA-3072 signed binary file.
+
+This plugin supports the following protocol ID:
+
+* com.ti.dmc FIXME
```
Comment (superm1) [inline plugins/ti-dmc/README.md:13]: left a FIXME here

File: plugins/ti-dmc/README.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Texas Instruments TPS6598x
```
Comment (superm1) [inline plugins/ti-dmc/README.md:1]: Shouldn't this have DMC in it?

File: plugins/ti-dmc/README.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+## Introduction
+
+The TPS65987D and TPS65988 are Type-C Port Switch / Power Delivery (PD) Controller devices.
```
Comment (superm1) [inline plugins/ti-dmc/README.md:5]: should the plugin be called `ti-tps6598x` instead?

File: plugins/ti-dmc/ti-dmc.quirk [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# TPS6598x
+[USB\VID_0451&PID_ACE1]
```
Comment (superm1) [inline plugins/ti-dmc/ti-dmc.quirk:2]: Is the PID going to be the same for all these devices or is it going to be changed on an OEM by OEM basis?

File: plugins/ti-dmc/meson.build [hunk 5]
```diff
@@ -0,0 +1,16 @@
+if gusb.found()
+cargs = ['-DG_LOG_DOMAIN="FuPluginTiDmc"']
+
+plugin_quirks += files('ti-dmc.quirk')
+plugin_builtins += static_library('fu_plugin_ti_dmc',
+  sources: [
+    'fu-ti-dmc-device.c',
+    'fu-ti-dmc-firmware.c',
```
Comment (superm1) [inline plugins/ti-dmc/meson.build:8]: don't you want this as part of `# fuzzing`?

File: plugins/ti-dmc/fu-ti-dmc-device.c [hunk 6]
```diff
@@ -0,0 +1,737 @@
...
+}
+
+static void
+fu_ti_dmc_device_init(FuTiDmcDevice *self)
+{
+	fu_device_add_protocol(FU_DEVICE(self), "com.ti.dmc");
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_PLAIN);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_SIGNED_PAYLOAD);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_DUAL_IMAGE);
```
Comment (superm1) [inline plugins/ti-dmc/fu-ti-dmc-device.c:726]: Being dual image, I would think this is _probably_ also self recovery.  Is that the case?

File: plugins/ti-dmc/ti-dmc.quirk (hunk 4 above)
Comment (hughsie) [reply plugins/ti-dmc/ti-dmc.quirk:2]: Urgh, you hit a nerve. They all have the same VID/PID with different firmware streams. I'm talking to TI about how to load the pubkey so we can add an extra instance ID.

File: plugins/ti-dmc/meson.build (hunk 5 above)
Comment (hughsie) [reply plugins/ti-dmc/meson.build:8]: I don't _think_ so. It's really just three blobs cat'd together, and I'm not sure it adds anything more than what we unit test `fu_bytes_new_offset()` already.

File: plugins/ti-dmc/fu-ti-dmc-device.c (hunk 6 above)
Comment (hughsie) [reply plugins/ti-dmc/fu-ti-dmc-device.c:726]: I'll ask, but I suspect so too.

**Comment by hughsie** [comment]
@anantgole-ti @d4s @mend0za I've deleted the private fork and we can continue discussion here. All review most welcome.

File: plugins/ti-dmc/fu-ti-dmc-device.c [hunk 7]
```diff
@@ -0,0 +1,737 @@
...
+typedef enum {
+	TI_DMC_SFWS_HASH_MATCH_NO_RULE_SATISFIED = 0,
+	TI_DMC_SFWS_HASH_MATCH_PROD_KEY_HASH_MATCH = 1,
+	TI_DMC_SFWS_HASH_MATCH_ENGR_KEY_HASH_MATCH = 2,
+	TI_DMC_SFWS_HASH_MATCH_ENGR_KEY_UID_OVERRIDE_MATCH = 3,
+	TI_DMC_SFWS_HASH_MATCH_UNKNOWN_HASH_MATCH_RULE_1 = 4,
+	TI_DMC_SFWS_HASH_MATCH_UNKNOWN_HASH_MATCH_RULE_2 = 5,
+	TI_DMC_SFWS_HASH_MATCH_UNKNOWN_HASH_MATCH_RULE_3 = 6,
+	TI_DMC_SFWS_HASH_MATCH_UNKNOWN_HASH_MATCH_RULE_4 = 7,
+} SFWsHashMatchRule;
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/ti-dmc/fu-ti-dmc-device.c:81]: All changed to #define

File: plugins/ti-dmc/fu-ti-dmc-device.c [hunk 8]
```diff
@@ -0,0 +1,737 @@
...
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 19, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 44, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 44, NULL);
+
+	/* get payload image */
+	fw_payload = fu_firmware_get_image_by_id_bytes(firmware, FU_FIRMWARE_ID_PAYLOAD, error);
+	if (fw_payload == NULL)
```
Comment (d4s) [inline plugins/ti-dmc/fu-ti-dmc-device.c:654]: Hm... here I have `NULL` while trying to use `install-blob` command with FWs I have.

Comment (hughsie) [reply plugins/ti-dmc/fu-ti-dmc-device.c:654]: oopa, can you try with the commit I just pushed -- we're not setting the firmware type on the device.

Comment (hughsie) [reply plugins/ti-dmc/fu-ti-dmc-device.c:647]: > and segfault after

Can you try the new commit -- I'm forgetting to create the g_byte_array... :/

Comment (d4s) [reply plugins/ti-dmc/fu-ti-dmc-device.c:647]: Sure, now have `response 0x7 but requested 0x40`.
on it.

Comment (hughsie) [reply plugins/ti-dmc/fu-ti-dmc-device.c:647]: So in the original code, read_reg says it always returns 64 bytes. Maybe the test should be `if (buf->data[0] < length)`? I think fwupd is being more strict that the original code.

Comment (d4s) [reply plugins/ti-dmc/fu-ti-dmc-device.c:647]: next step `SFWs failed`

Comment (hughsie) [reply plugins/ti-dmc/fu-ti-dmc-device.c:647]: Is it a bit weird that "SFWs with signature" and "SFWs with pubkey" both call fu_ti_dmc_device_write_sfws_chunks?

Comment (hughsie) [reply plugins/ti-dmc/fu-ti-dmc-device.c:722]: Maybe not so weird. I tried myself and pushed what we have so far, and a few lines to add proper error messages. Now we have:

    failed to write SFWs with signature: failed to write chunk 0: SFWs failed, got sfwd-not-run-or-no-key-exists [0x06]

you see the same?

File: plugins/ti-dmc/ti-dmc.quirk (hunk 4 above)
Comment (superm1) [reply plugins/ti-dmc/ti-dmc.quirk:2]: OK; good.  Besides the pubkey, is there an OEM specific string that can be pulled out for the device name?

File: plugins/ti-dmc/fu-ti-dmc-device.c (hunk 8 above)
Comment (hughsie) [reply plugins/ti-dmc/fu-ti-dmc-device.c:722]: Also, debugging this a bit more I see:

	36 5a 02 20 ff 00 00 00 00 80 30 5a 02 20 79 42 5a 02 20 ff 00 00 00 00 01 00 00 ef 00 00 00 00 
	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	15:17:55.489 FuPluginTiDmc        write@0x1e:     53 46 57 64 
	15:17:55.490 FuPluginTiDmc        read@0x1e:      04 53 46 57 64 
	15:17:55.591 FuPluginTiDmc        read@0x1e:      04 00 00 00 00 
	15:17:55.592 FuPluginTiDmc        read@0x1f:      40 80 
	15:17:55.592 FuPluginTiDmc        more-data-expected: 1
	15:17:55.692 FuPluginTiDmc        write@0x1f:
	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	15:17:55.699 FuPluginTiDmc        write@0x1e:     53 46 57 64 
	15:17:55.700 FuPluginTiDmc        read@0x1e:      04 53 46 57 64 
	15:17:55.801 FuPluginTiDmc        read@0x1e:      04 00 00 00 00 
	15:17:55.802 FuPluginTiDmc        read@0x1f:      40 80 
	15:17:55.802 FuPluginTiDmc        more-data-expected: 1
	15:17:55.902 FuPluginTiDmc        write@0x1f:
	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ff ff ff ff 
	15:17:55.909 FuPluginTiDmc        write@0x1e:     53 46 57 64 
	15:17:55.910 FuPluginTiDmc        read@0x1e:      04 53 46 57 64 
	15:17:56.011 FuPluginTiDmc        read@0x1e:      04 00 00 00 00 
	15:17:56.012 FuPluginTiDmc        read@0x1f:      40 80 
	15:17:56.012 FuPluginTiDmc        more-data-expected: 1

i.e. we're writing all the SFWd blocks, and even padding out the last block with `0xFF` like we expect. But we still get the `more-data-expected: 1` -- I did kinda expect that to be false for the last block.

File: plugins/ti-dmc/ti-dmc.quirk (hunk 4 above)
Comment (hughsie) [reply plugins/ti-dmc/ti-dmc.quirk:2]: Alas not, all `Texas Instruments, Inc. TPS DMC Family`

File: plugins/ti-dmc/fu-ti-dmc-device.c (hunk 8 above)
Comment (d4s) [reply plugins/ti-dmc/fu-ti-dmc-device.c:725]: not exactly.
In my case the data from common_update is written and after that I've got a failure:

```
14:34:40.327 FuPluginTiDmc        SFWd: 0x80, 0x97
14:34:40.428 FuPluginTiDmc        Write@0x1f:
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ff ff ff ff
14:34:40.433 FuPluginTiDmc        Write@0x1e:     53 46 57 64
14:34:40.434 FuPluginTiDmc        Read@0x1e:      04 53 46 57 64
14:34:40.534 FuPluginTiDmc        Read@0x1e:      04 00 00 00 00
14:34:40.536 FuPluginTiDmc        Read@0x1f:
       | 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f
-------+------------------------------------------------------------------------------------------------
0x0000 | 40 80 98 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0x0020 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ff ff ff
0x0040 | ff
14:34:40.536 FuPluginTiDmc        SFWd: 0x80, 0x98
14:34:40.636 FuProgress           child is at 664/750 step_max and parent done [../plugins/ti-dmc/fu-ti-dmc-device.c:642]
4) ../src/fu-tool.c:892 (2/3)
3) ../src/fu-engine.c:3454 (1/3)
2) ../plugins/ti-dmc/fu-ti-dmc-device.c:708 (1/4)
1) ../plugins/ti-dmc/fu-ti-dmc-device.c:642 (1/4)
0) ../plugins/ti-dmc/fu-ti-dmc-device.c:559 (664/750)
  
14:34:40.637 FuPluginTiDmc        Write@0x1f:
27 a8 fa 49 46 1f 07 ae 32 75 41 3a 0f d1 88 1d 34 54 05 58 01 7d f3 b2 44 ec e0 43 ad 7f 80 f5
57 3e 46 c7 3a 14 95 87 98 2d 32 0b 78 67 b8 50 5d 01 b8 32 6c cd 99 ee b5 39 c4 69 af b3 a6 86
14:34:40.641 FuPluginTiDmc        Write@0x1e:     53 46 57 73
14:34:40.642 FuPluginTiDmc        Read@0x1e:      04 00 00 00 00
14:34:40.644 FuPluginTiDmc        Read@0x1f:
       | 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f
-------+------------------------------------------------------------------------------------------------
0x0000 | 40 06 00 00 00 46 1f 07 ae 32 75 41 3a 0f d1 88 1d 34 54 05 58 01 7d f3 b2 44 ec e0 43 ad 7f 80
0x0020 | f5 57 3e 46 c7 3a 14 95 87 98 2d 32 0b 78 67 b8 50 5d 01 b8 32 6c cd 99 ee b5 39 c4 69 af b3 a6
0x0040 | 86
14:34:40.644 FuPluginTiDmc        SFWs: 0x6, 0x0
14:34:40.646 FuPlugin             attach(ti_dmc)
14:34:40.646 FuDeviceList         no replug or re-enumerate required
14:34:40.646 FuEngine             cleanup -> FuTiDmcDevice:

14:34:40.646 FuDeviceList         no replug or re-enumerate required
SFWs failed, got 0x06
```
Now checking the original workflow in python to compare from scratch.

Comment (hughsie) [reply plugins/ti-dmc/fu-ti-dmc-device.c:725]: On a [risky?] hunch I disabled the ignoring of duplicate chunks, and how I get `more-data-expected: 0` and:

15:31:43.430 FuPluginTiDmc        read@0x1e:      04 00 00 00 00
	15:31:43.430 FuPluginTiDmc        read@0x1f:      40 00 0c 36 00 00 00 00 00 00 00
	15:31:43.430 FuPluginTiDmc        more-data-expected: 0
	15:31:43.430 FuPluginTiDmc        signature-data-block: 12
	15:31:43.430 FuPluginTiDmc        prod-key-present: 2
	15:31:43.430 FuPluginTiDmc        engr-key-present: 4
	15:31:43.430 FuPluginTiDmc        new-flash-region: 16

...which I *think* looks like a successful flash!

The device seems to reboot after calling fu_ti_dmc_device_reset_hard() -- which is maybe expected!

Comment (d4s) [reply plugins/ti-dmc/fu-ti-dmc-device.c:725]: argh, I didn't pull last changes -- that's why....

File: plugins/ti-dmc/README.md (hunk 3 above)
Comment (hughsie) [reply plugins/ti-dmc/README.md:5]: Agreed, renamed.

File: plugins/ti-dmc/README.md (hunk 2 above)
Comment (hughsie) [reply plugins/ti-dmc/README.md:1]: Fixed.

File: plugins/ti-dmc/README.md (hunk 1 above)
Comment (hughsie) [reply plugins/ti-dmc/README.md:13]: Fixed with the rename.

File: plugins/ti-dmc/fu-ti-dmc-device.c (hunk 6 above)
Comment (hughsie) [reply plugins/ti-dmc/fu-ti-dmc-device.c:798]: It is, fixed.

File: plugins/ti-tps6598x/fu-ti-tps6598x-pd-device.c [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+#include "fu-ti-tps6598x-pd-device.h"
+
+struct _FuTiTps6598xPdDevice {
+	FuDevice parent_instance;
+	guint8 target;
+};
+
+G_DEFINE_TYPE(FuTiTps6598xPdDevice, fu_ti_tps6598x_pd_device, FU_TYPE_DEVICE)
+
+// FIXME move to common code
```
Comment (d4s) [inline plugins/ti-tps6598x/fu-ti-tps6598x-pd-device.c:21]: missed FIXME?

File: plugins/ti-tps6598x/fu-ti-tps6598x-pd-device.c [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+		return FALSE;
+	config = fu_byte_array_to_string(buf);
+	fu_device_add_instance_strup(device, "CONFIG", config);
+
+	/* success */
+	return fu_device_build_instance_id(device, error, "USB", "VID", "PID", "CONFIG", NULL);
+}
+
+static gboolean
+fu_ti_tps6598x_pd_device_byte_array_is_interesting(GByteArray *buf)
```
Comment (d4s) [inline plugins/ti-tps6598x/fu-ti-tps6598x-pd-device.c:92]: Is it really a good name for this function? ;)

Comment (hughsie) [reply plugins/ti-tps6598x/fu-ti-tps6598x-pd-device.c:92]: `fu_ti_tps6598x_pd_device_byte_array_is_zero` perhaps?

File: plugins/ti-tps6598x/fu-ti-tps6598x-pd-device.c (hunk 9 above)
Comment (hughsie) [reply plugins/ti-tps6598x/fu-ti-tps6598x-pd-device.c:21]: Done in https://github.com/fwupd/fwupd/pull/5327 -- will remove on rebase.

File: plugins/ti-tps6598x/fu-ti-tps6598x-pd-device.c (hunk 10 above)
Comment (d4s) [reply plugins/ti-tps6598x/fu-ti-tps6598x-pd-device.c:92]: Sounds good.

**Comment by d4s** [comment]
LGTM, however the `fwupdtool` stuck in my tests -- the previous variant works well.
I did the rebase against main.
Attached the log from the single test.
[write-8-TS4-combind-122021-secure-Cus-08.log](https://github.com/fwupd/fwupd/files/10223684/write-8-TS4-combind-122021-secure-Cus-08.log)

File: plugins/ti-tps6598x/fu-ti-tps6598x-pd-device.c (hunk 10 above)
Comment (hughsie) [reply plugins/ti-tps6598x/fu-ti-tps6598x-pd-device.c:92]: Fixed.

**Review by d4s — DISMISSED** [review]

**Pull request merged**